geth/api/backend.go

```diff
@@ -77,120 +78,97 @@ func (m *StatusBackend) IsNodeRunning() bool {
+		}
+	}()
+	err = m.nodeManager.StartNode(config)
 	if err != nil {
-		return nil, err
+		switch err.(type) {
+		case node.RPCClientError:
+			err = node.ErrRPCClient
+		case node.EthNodeError:
+			err = fmt.Errorf("%v: %v", node.ErrNodeStartFailure, err)
+		}
+		signal.Send(signal.Envelope{
+			Type: signal.EventNodeCrashed,
+			Event: signal.NodeCrashEvent{
+				Error: err,
+			},
+		})
+		return err
 	}
-
-	m.nodeReady = make(chan struct{}, 1)
-	go m.onNodeStart(nodeStarted, m.nodeReady) // waits on nodeStarted, writes to backendReady
-
-	return m.nodeReady, err
+	// not sure if we should keep both, onNodeStart is so simple, that it doesn't
+	signal.Send(signal.Envelope{Type: signal.EventNodeStarted})
+	m.onNodeStart()
+	signal.Send(signal.Envelope{Type: signal.EventNodeReady})
+	return nil
 }
 
 // onNodeStart does everything required to prepare backend
-func (m *StatusBackend) onNodeStart(nodeStarted <-chan struct{}, backendReady chan struct{}) {
-	<-nodeStarted
-
+func (m *StatusBackend) onNodeStart() {
 	// tx queue manager should be started after node is started, it depends
 	// on rpc client being created
 	m.txQueueManager.Start()
-
 	if err := m.registerHandlers(); err != nil {
 		log.Error("Handler registration failed", "err", err)
 	}
-
 	if err := m.accountManager.ReSelectAccount(); err != nil {
 		log.Error("Reselect account failed", "err", err)
 	}
 	log.Info("Account reselected")
-
-	close(backendReady)
-	signal.Send(signal.Envelope{
-		Type:  signal.EventNodeReady,
-		Event: struct{}{},
-	})
 }
 
 // StopNode stop Status node. Stopped node cannot be resumed.
-func (m *StatusBackend) StopNode() (<-chan struct{}, error) {
-	m.Lock()
-	defer m.Unlock()
+func (m *StatusBackend) StopNode() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.stopNode()
+}
 
-	if m.nodeReady == nil {
-		return nil, node.ErrNoRunningNode
+func (m *StatusBackend) stopNode() error {
+	if !m.IsNodeRunning() {
+		return node.ErrNoRunningNode
 	}
-	<-m.nodeReady
-
 	m.txQueueManager.Stop()
 	m.jailManager.Stop()
-
-	nodeStopped, err := m.nodeManager.StopNode()
-	if err != nil {
-		return nil, err
-	}
-
-	backendStopped := make(chan struct{}, 1)
-	go func() {
-		<-nodeStopped
-		m.Lock()
-		m.nodeReady = nil
-		m.Unlock()
-		close(backendStopped)
-	}()
-
-	return backendStopped, nil
+	return m.nodeManager.StopNode()
 }
 
 // RestartNode restart running Status node, fails if node is not running
-func (m *StatusBackend) RestartNode() (<-chan struct{}, error) {
-	m.Lock()
-	defer m.Unlock()
-
-	if m.nodeReady == nil {
-		return nil, node.ErrNoRunningNode
-	}
-	<-m.nodeReady
-
-	nodeRestarted, err := m.nodeManager.RestartNode()
-	if err != nil {
-		return nil, err
-	}
-
-	m.nodeReady = make(chan struct{}, 1)
-	go m.onNodeStart(nodeRestarted, m.nodeReady) // waits on nodeRestarted, writes to backendReady
-
-	return m.nodeReady, err
+func (m *StatusBackend) RestartNode() error {
+	return m.nodeManager.RestartNode()
```

Comment:
we don't need to lock/unlock here, or do we?

Reply:
nodeManager is thread safe already

Reply:
cool, that might be nice to point out (in comment, maybe)?
or just add mutex here for consistency (restart is already a heavy operation, so probably impact will be tiny).
From the first sight this method looks weird, because it seems like it's a `StatusBackend`'s responsibility to make everything thread-safe.

Reply:
Now `StatusBackend` kinda "knows" that this method of `NodeManager` is thread-safe, which is, IMO, an implementation details.

(hope I don't sound like this OOP guru woowoo now ;)